jgrapht-core/src/test/java/org/jgrapht/graph/AsSynchronizedGraphTest.java

```diff
@@ -0,0 +1,476 @@
+        g.removeVertex(vertices.get(1));
+        assertEquals(9, g.vertexSet().size());
+        assertEquals(9, iteratorCnt(g.vertexSet().iterator()));
+        assertEquals(81, g.edgeSet().size());
+        assertEquals(81, iteratorCnt(g.edgeSet().iterator()));
+        assertEquals(9, g.incomingEdgesOf(vertices.get(0)).size());
+        assertEquals(9, g.outgoingEdgesOf(vertices.get(0)).size());
+        assertEquals(17, g.edgesOf(vertices.get(0)).size());
+    }
+
+    @Test
+    public void testOthers()
+    {
+        g = new AsSynchronizedGraph<>(new Pseudograph<>(DefaultEdge.class),
+                new SynchronizedGraphParams().cacheDisable());
+        Set<Integer> vertSet = g.vertexSet();
+        Set<DefaultEdge> edgeSet = g.edgeSet();
+        g.addVertex(1);
+        g.addVertex(2);
+        assertEquals(2, vertSet.size());
+        assertEquals(2, iteratorCnt(vertSet.iterator()));
+        g.addVertex(3);
+        g.addVertex(4);
+        assertEquals(4, vertSet.size());
+        assertEquals(4, iteratorCnt(vertSet.iterator()));
+        assertEquals(0, edgeSet.size());
+        assertEquals(0, iteratorCnt(edgeSet.iterator()));
+
+        g.addEdge(1,2);
+        assertEquals(1, g.edgesOf(2).size());
+        assertEquals(1, iteratorCnt(g.edgesOf(2).iterator()));
+        assertEquals(1, g.outgoingEdgesOf(1).size());
+        assertEquals(1, g.incomingEdgesOf(2).size());
+
+        g.addEdge(2,3);
+        assertEquals(2, edgeSet.size());
+        assertEquals(2, iteratorCnt(edgeSet.iterator()));
+        assertEquals(2, g.edgesOf(2).size());
+        assertEquals(2, iteratorCnt(g.edgesOf(2).iterator()));
+        assertEquals(2, g.outgoingEdgesOf(2).size());
+        assertEquals(2, g.incomingEdgesOf(2).size());
+        assertFalse(g.isCacheEnabled());
+        g.setCache(true);
+        assertTrue(g.isCacheEnabled());
+        g.addEdge(2, 4);
+        assertEquals(3, g.edgesOf(2).size());
+        assertEquals(3, iteratorCnt(g.edgesOf(2).iterator()));
+        assertEquals(3, g.outgoingEdgesOf(2).size());
+        assertEquals(3, g.incomingEdgesOf(2).size());
+        g.addEdge(2,2);
+        assertEquals(4, g.edgesOf(2).size());
+        assertEquals(4, iteratorCnt(g.edgesOf(2).iterator()));
+        assertEquals(4, g.outgoingEdgesOf(2).size());
+        assertEquals(4, g.incomingEdgesOf(2).size());
+
+        g.removeVertex(3);
+        assertEquals(3, vertSet.size());
+        assertEquals(3, iteratorCnt(vertSet.iterator()));
+        assertEquals(3, edgeSet.size());
+        assertEquals(3, iteratorCnt(edgeSet.iterator()));
+        assertEquals(3, g.edgeSet().size());
+        assertEquals(3, g.incomingEdgesOf(2).size());
+        assertEquals(3, g.outgoingEdgesOf(2).size());
+    }
+
+    private ArrayList<Order> order1;
+    private ArrayList<Order> order2;
+    private ArrayList<Order> order3;
+    private ArrayList<Order> order4;
+    volatile private Vector<ArrayList<Order> > ordersList;
```

Comment:
Is volatile really necessary?  The content of the vector is volatile, but the reference to the vector object itself is not.

Reply:
Sorry, I am concerned about instruction reordering. But it seems that the worry is superfluous.